sycl-fusion/passes/kernel-fusion/Builtins.cpp

```diff
@@ -442,59 +425,99 @@ Value *jit_compiler::getGlobalLinearID(IRBuilderBase &Builder,
   return createSPIRVCall(Builder, GetGlobalLinearIDName, {});
 }
 
-Function *jit_compiler::remapBuiltins(Function *F, const NDRange &SrcNDRange,
+static bool isIndexSpaceGetterBuiltin(Function *F) {
+  constexpr std::array<StringLiteral, NumBuiltinsToRemap> BuiltinNames{
+      GetGlobalSizeName,    GetGroupIDName,   GetGlobalOffsetName,
+      GetNumWorkGroupsName, GetLocalSizeName, GetLocalIDName,
+      GetGlobalIDName};
+  const auto Name = F->getName();
+  const auto *Iter = llvm::lower_bound(BuiltinNames, Name);
+  return Iter != BuiltinNames.end() && *Iter == Name;
+}
+
+static bool isSafeToNotRemapBuiltin(Function *F) {
+  constexpr std::size_t NumUnsafeBuiltins{8};
+  // SPIRV builtins with kernel capabilities in alphabetical order.
+  constexpr std::array<StringLiteral, NumUnsafeBuiltins> UnsafeBuiltIns{
+      "EnqueuedWorkgroupSize",
+      "NumEnqueuedSubgroups",
+      "NumSubgroups",
+      "SubgroupId",
+      "SubgroupLocalInvocationId",
+      "SubgroupMaxSize",
+      "SubgroupSize",
+      "WorkDim"};
+  constexpr StringLiteral SPIRVBuiltinNamespace{"spirv"};
+  constexpr StringLiteral SPIRVBuiltinPrefix{"BuiltIn"};
+
+  auto Name = F->getName();
+  if (!(Name.contains(SPIRVBuiltinNamespace) &&
+        Name.contains(SPIRVBuiltinPrefix))) {
+    return true;
+  }
+  // Drop "spirv" namespace name and "BuiltIn" prefix.
+  Name = Name.drop_front(Name.find(SPIRVBuiltinPrefix) +
+                         SPIRVBuiltinPrefix.size());
+  // Check that Name does not start with any name in UnsafeBuiltIns
+  const auto *Iter = llvm::upper_bound(UnsafeBuiltIns, Name);
```

Comment:
Any particular reason to use `llvm::upper_bound` and not `std::upper_bound` here?

Reply:
None, just wanted to use ranges instead of iterators, but it's true we'd have to update *when* (I'm assuming it'll happen) these are dropped in favor of the new `std` signatures.